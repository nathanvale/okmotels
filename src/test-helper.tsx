import React from 'react'

import {render, RenderResult, RenderOptions} from 'react-testing-library'
import {ThemeProvider} from 'styled-components'
import {Omit} from './helper'

export const mock = <T extends {}, K extends keyof T>(
  object: T,
  property: K,
  value: T[K],
): void => {
  Object.defineProperty(object, property, {get: () => value})
}

export const renderWithTheme = (
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  theme: any,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ui: React.ReactElement<any>,
  options?: Omit<RenderOptions, 'queries'>,
): RenderResult =>
  render(<ThemeProvider theme={theme}>{ui}</ThemeProvider>, options)
