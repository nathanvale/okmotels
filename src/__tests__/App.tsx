import React from 'react'
import {
  render,
  fireEvent,
  RenderResult,
  waitForDomChange,
} from 'react-testing-library'
import {
  createHistory,
  createMemorySource,
  LocationProvider,
  History,
} from '@reach/router'
import * as gatsby from 'gatsby'
import App from '../pages/app'
import {mock} from '../test-helper'

jest.mock('gatsby')

function renderWithRouter(
  ui: React.ReactNode,
  {route = '/', history = createHistory(createMemorySource(route))} = {},
): RenderResult & {history: History} {
  return {
    ...render(<LocationProvider history={history}>{ui}</LocationProvider>),
    history,
  }
}

test.skip('login page redirects to profile page on correct authentication', async () => {
  const {
    getByLabelText,
    getByTestId,
    getByText,
    history: {navigate},
  } = renderWithRouter(<App />, {
    route: '/app/login',
  })
  mock(gatsby, 'navigate', jest.fn(navigate))
  expect(getByTestId('login')).toMatchSnapshot()
  fireEvent.change(getByLabelText(/Username/i), {target: {value: 'nathanvale'}})
  fireEvent.change(getByLabelText(/Password/i), {target: {value: 'qwerty'}})
  fireEvent.click(getByText(/Log In/i))
  await waitForDomChange()
  expect(getByTestId('profile')).toMatchSnapshot()
})
