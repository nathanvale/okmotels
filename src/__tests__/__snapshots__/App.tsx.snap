// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login page redirects to profile page on correct authentication 1`] = `
<section
  class="Container__Wrapper-koxfXz bHtbAd"
  data-testid="login"
>
  <h1
    class="PageTitle__Title-euxURV bAxgis"
  >
    Login
  </h1>
  <form
    class="FormElements__Form-gUrTuf cFQswl"
    method="post"
  >
    <label
      for="username"
      hidden=""
    >
      Username
    </label>
    <input
      class="FormElements__Input-kwjIXe extkka"
      id="username"
      name="username"
      placeholder="Username"
      type="text"
    />
    <label
      for="password"
      hidden=""
    >
      Password
    </label>
    <input
      class="FormElements__Input-kwjIXe extkka"
      id="password"
      name="password"
      placeholder="Password"
      type="password"
    />
    <input
      class="FormElements__Submit-csJEoT brxrRQ"
      type="submit"
      value="Log In"
    />
  </form>
</section>
`;

exports[`login page redirects to profile page on correct authentication 2`] = `
<section
  class="Container__Wrapper-koxfXz bHtbAd"
  data-testid="profile"
>
  <h1>
    Your profile
  </h1>
  <ul>
    <li>
      Name: 
      Nathan
    </li>
    <li>
      E-mail: 
      hi@nathanvale.com
    </li>
  </ul>
</section>
`;
